common/raw_hashtable.h

```diff
@@ -0,0 +1,1206 @@
+//
+// The methods available to user-facing hashtable types are `protected`, and
+// where relevant named with an `Impl` suffix. The suffix naming ensures types
+// don't `using` in these low-level APIs but declare their own and implement
+// them by forwarding to these APIs. We don't want users to have to read these
+// implementation details to understand their container's API.
+//
+// Many method are `private` and used purely by other parts of the raw hashtable
+// implementation. Where needed, friendship is used to access those.
+template <typename InputKeyT, typename InputValueT>
+class BaseImpl {
+ protected:
+  using KeyT = InputKeyT;
+  using ValueT = InputValueT;
+  using ViewImplT = ViewImpl<KeyT, ValueT>;
+  using EntryT = typename ViewImplT::EntryT;
+
+  BaseImpl(int small_size, Storage* small_storage) : small_size_(small_size) {
+    CARBON_CHECK(small_size >= 0);
+    Construct(small_storage);
+  }
+
+  ~BaseImpl();
+
+  // NOLINTNEXTLINE(google-explicit-constructor): Designed to implicitly decay.
+  operator ViewImplT() const { return view_impl(); }
+
+  auto view_impl() const -> ViewImplT { return impl_view_; }
+
+  template <typename LookupKeyT>
+  auto InsertImpl(LookupKeyT lookup_key) -> std::pair<EntryT*, bool>;
+  template <typename LookupKeyT>
+  auto EraseImpl(LookupKeyT lookup_key) -> bool;
+  auto ClearImpl() -> void;
+
+ private:
+  template <typename InputBaseT, ssize_t SmallSize>
+  friend class TableImpl;
+
+  static constexpr ssize_t Alignment = std::max<ssize_t>(
+      {alignof(MetadataGroup), alignof(StorageEntry<KeyT, ValueT>)});
+
+  static constexpr auto Size(ssize_t size) -> ssize_t {
+    return ViewImplT::EntriesOffset(size) + sizeof(EntryT) * size;
+  }
+  static auto Allocate(ssize_t size) -> Storage*;
+  static auto Deallocate(Storage* storage, ssize_t size) -> void;
+
+  auto growth_budget() const -> ssize_t { return growth_budget_; }
+  auto size() const -> ssize_t { return impl_view_.size_; }
+  auto size() -> ssize_t& { return impl_view_.size_; }
+  auto storage() const -> Storage* { return impl_view_.storage_; }
+  auto storage() -> Storage*& { return impl_view_.storage_; }
+  auto metadata() const -> uint8_t* { return impl_view_.metadata(); }
+  auto entries() const -> EntryT* { return impl_view_.entries(); }
+  auto small_size() const -> ssize_t {
+    return static_cast<unsigned>(small_size_);
+  }
+  auto is_small() const -> bool { return size() <= small_size(); }
+
+  auto Construct(Storage* small_storage) -> void;
+  auto CopyFrom(const BaseImpl& arg) -> void;
+  auto MoveFrom(BaseImpl&& arg) -> void;
+  auto Destroy() -> void;
+
+  template <typename LookupKeyT>
+  auto InsertIntoEmpty(LookupKeyT lookup_key) -> EntryT*;
+
+  static auto ComputeNewSize(ssize_t old_size) -> ssize_t;
+  static auto GrowthThresholdForSize(ssize_t size) -> ssize_t;
+
+  template <typename LookupKeyT>
+  auto GrowAndInsert(LookupKeyT lookup_key) -> EntryT*;
+
+  ViewImplT impl_view_;
```

Comment:
Any reason for these not to match?
```suggestion
  ViewImplT view_impl_;
```

Reply:
No, I forgot to correct them to match, thanks.